packages/models-library/src/models_library/utils/string_substitution.py

```diff
@@ -0,0 +1,102 @@
+""" Utils for substitutions in string templates
+
+"""
+import re
+from collections import UserDict
+from string import Template
+from typing import Any
+
+
+def upgrade_identifier(legacy_id):
+    legacy_id = re.sub(r"[{}%]", "", legacy_id)
+    legacy_id = re.sub(r"[.-]", "_", legacy_id)
+    legacy_id = legacy_id.upper()
+    if not legacy_id.startswith("OSPARC_ENVIRONMENT_"):
+        legacy_id = f"OSPARC_ENVIRONMENT_{legacy_id}"
+    return legacy_id
+
+
+LEGACY_FULL_IDENTIFIER_PATTERN = re.compile(r"%{1,2}([_a-z][_a-z0-9\.\-]*)%{1,2}")
+
+
+def substitute_all_legacy_identifiers(text: str):
+    """Substitutes all legacy identifiers found in the text by the new format expected in TemplateText
+
+    For instance:  '%%this-identifier%%' will be substituted by '$OSPARC_ENVIRONMENT_THIS_IDENTIFIER'
+    """
+
+    def _upgrade(match):
+        legacy_id = match.group(1)
+        legacy_id = upgrade_identifier(legacy_id)
```

Comment:
so if I get that right, you pass the old syntax through that function, which converts to the new syntax. nice!
ist ```upgrade_identifier``` a private method?

Reply:
yes.
no, is public